src/stream/src/executor/source/source_executor.rs

```diff
@@ -490,19 +505,61 @@ impl<S: StateStore> SourceExecutor<S> {
 
         let recover_state: ConnectorState = (!boot_state.is_empty()).then_some(boot_state);
         tracing::debug!(state = ?recover_state, "start with state");
-        let (source_chunk_reader, latest_splits) = self
-            .build_stream_source_reader(
-                &source_desc,
-                recover_state,
-                // For shared source, we start from latest and let the downstream SourceBackfillExecutors to read historical data.
-                // It's highly probable that the work of scanning historical data cannot be shared,
-                // so don't waste work on it.
-                // For more details, see https://github.com/risingwavelabs/risingwave/issues/16576#issuecomment-2095413297
-                // Note that shared CDC source is special. It already starts from latest.
-                self.is_shared_non_cdc && is_uninitialized,
+
+        let mut barrier_stream = barrier_to_message_stream(barrier_receiver).boxed();
+        let mut reader_and_splits: Option<(BoxChunkSourceStream, Option<Vec<SplitImpl>>)> = None;
+        let seek_to_latest = self.is_shared_non_cdc && is_uninitialized;
+        let source_reader = source_desc.source.clone();
+        let (column_ids, source_ctx) = self.prepare_source_stream_build(&source_desc);
+        let source_ctx = Arc::new(source_ctx);
+        let mut build_source_stream_fut = Box::pin(async move {
+            let backoff = get_infinite_backoff_strategy();
+            tokio_retry::Retry::spawn(backoff, || async {
+                match source_reader
+                    .build_stream(
+                        recover_state.clone(),
+                        column_ids.clone(),
+                        source_ctx.clone(),
+                        seek_to_latest,
+                    )
+                    .await {
+                    Ok((stream, latest_splits)) => Ok((stream, latest_splits)),
+                    Err(e) => {
+                        tracing::warn!(error = %e.as_report(), "failed to build source stream, retrying...");
+                        Err(e)
+                    },
+                }
+            })
+            .instrument(tracing::info_span!("build_source_stream_with_retry"))
+            .await
+            .expect("Retry build source stream until success.")
+        });
+
+        // loop to create source stream until success
+        loop {
+            if let Some(barrier) = build_source_stream_and_poll_barrier(
```

Comment:
@xxchan @tabVersion PTAL this part for source executor

Reply:
The changes here apply to all sources. Does it mean that prior to this PR when upstream system is not available, the cluster will fail to start up?

Reply:
I guess it's `Barrier::Resume` not handled correctly for replace table.